### ISSUE TITLE
no-unnecessary-type-assertion false positive when casting never to any

### ISSUE THREAD

**Issue by adidahiya** [palantir/tslint#4091]
### Bug Report

- __TSLint version__: 5.11.0
- __TypeScript version__: 2.8.4
- __Running TSLint via__: CLI

#### TypeScript code being linted

```ts
const foo: { type: "a" | "b" } = { type: "a" };

if (foo.type === "a") {
  // ...
} else if (foo.type === "b") {
  // ...
} else {
  console.log(`Got unexpected foo: "${(foo as any).type}"`);
}
```

#### Actual behavior

```
no-unnecessary-type-assertion  This assertion is unnecessary since it does not change the type of the expression.
```

#### Expected behavior

No lint failure

**Comment by JoshuaKGoldberg** [comment]
### 💀 _**It's time!**_ 💀

**TSLint is deprecated** and no longer accepting pull requests other than security fixes. See #4534. ☠️ 
We recommend you instead use **[typescript-eslint](https://typescript-eslint.io)** to lint your TypeScript code with **ESLint**. ✅

👋 It was a pleasure open sourcing with you!

**Comment by JoshuaKGoldberg** [comment]
🤖 Beep boop! 👉  **TSLint is deprecated** 👈 _(#4534)_  and you should **switch to [typescript-eslint](https://typescript-eslint.io)**! 🤖 

🔒 This issue is being locked to prevent further unnecessary discussions. Thank you! 👋
